x-pack/solutions/observability/plugins/apm/common/service_map/get_service_map_nodes.ts

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { sortBy } from 'lodash';
+import type { ServiceAnomaliesResponse } from '../../server/routes/service_map/get_service_anomalies';
+import {
+  SERVICE_NAME,
+  SPAN_DESTINATION_SERVICE_RESOURCE,
+  SPAN_TYPE,
+  SPAN_SUBTYPE,
+} from '../es_fields/apm';
+import type { ExitSpanDestination, ServicesResponse } from './types';
+import type {
+  Connection,
+  ConnectionNode,
+  ServiceConnectionNode,
+  ExternalConnectionNode,
+  ConnectionElement,
+  ConnectionEdge,
+  ServiceMapConnections,
+  GroupResourceNodesResponse,
+} from './types';
+
+import { groupResourceNodes } from './group_resource_nodes';
+import { getConnectionId, isExitSpan } from './utils';
+
+const FORBIDDEN_SERVICE_NAMES = ['constructor'];
+
+function addMessagingConnections(
+  connections: Connection[],
+  destinationServices: ExitSpanDestination[]
+): Connection[] {
+  const serviceMap = new Map(
+    destinationServices.map(({ from, to }) => [from[SPAN_DESTINATION_SERVICE_RESOURCE], to])
+  );
+
+  const messagingConnections = connections.reduce<Connection[]>((acc, connection) => {
+    const destination = connection.destination;
+    if (isExitSpan(destination) && destination[SPAN_TYPE] === 'messaging') {
+      const matchedService = serviceMap.get(destination[SPAN_DESTINATION_SERVICE_RESOURCE]);
+      if (matchedService) {
+        acc.push({
+          source: destination,
+          destination: matchedService,
+        });
+      }
+    }
+    return acc;
+  }, []);
+
+  return [...connections, ...messagingConnections];
+}
+
+function getAllNodes(services: ServicesResponse[], connections: Connection[]) {
+  const allNodesMap = new Map<string, ConnectionNode>();
+
+  connections.forEach((connection) => {
+    const sourceId = connection.source.id;
+    const destinationId = connection.destination.id;
+    if (!allNodesMap.has(sourceId)) {
+      allNodesMap.set(sourceId, { ...connection.source, id: sourceId });
+    }
+    if (!allNodesMap.has(destinationId)) {
+      allNodesMap.set(destinationId, { ...connection.destination, id: destinationId });
+    }
+  });
+
+  // Derive the rest of the map nodes from the connections and add the services
+  // from the services data query
+  services
+    .filter((service) => !FORBIDDEN_SERVICE_NAMES.includes(service[SERVICE_NAME]))
```

Comment:
This `filter` can just be part of the `forEach` and avoids allocating an extra temporary array.

Reply:
Absolutely